internal/handlers/pg/pgdb/query_documents_test.go

```diff
@@ -0,0 +1,174 @@
+			collection:       collectionName + "_one",
+			documents:        []*types.Document{must.NotFail(types.NewDocument("id", "1"))},
+			docsPerIteration: []int{1},
+		},
+		{
+			name:       "two",
+			collection: collectionName + "_two",
+			documents: []*types.Document{
+				must.NotFail(types.NewDocument("id", "1")),
+				must.NotFail(types.NewDocument("id", "2")),
+			},
+			docsPerIteration: []int{2},
+		},
+		{
+			name:       "three",
+			collection: collectionName + "_three",
+			documents: []*types.Document{
+				must.NotFail(types.NewDocument("id", "1")),
+				must.NotFail(types.NewDocument("id", "2")),
+				must.NotFail(types.NewDocument("id", "3")),
+			},
+			docsPerIteration: []int{2, 1},
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			// t.Parallel()
+
+			for _, doc := range tc.documents {
+				require.NoError(t, pool.InsertDocument(ctx, dbName, tc.collection, doc))
+			}
+
+			fetchedChan, err := pool.QueryDocuments(context.Background(), dbName, tc.collection, "")
+			require.NoError(t, err)
+
+			iter := 0
+			for {
+				fetched, ok := <-fetchedChan
+				if !ok {
+					break
+				}
+
+				require.NoError(t, fetched.Err)
+				require.Len(t, fetched.Docs, tc.docsPerIteration[iter], "iteration %d", iter)
+				iter++
+			}
+			require.Equal(t, len(tc.docsPerIteration), iter)
+		})
+	}
+
+	// Special case: cancel context before reading from channel.
+	t.Run("cancel_context", func(t *testing.T) {
+		for i := 1; i <= pgdb.FetchedChannelCapacity*pgdb.FetchedSliceCapacity+1; i++ {
+			require.NoError(t, pool.InsertDocument(
+				ctx, dbName, collectionName+"_cancel",
+				must.NotFail(types.NewDocument("id", fmt.Sprintf("%d", i))),
+			))
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		fetchedChan, err := pool.QueryDocuments(ctx, dbName, collectionName+"_cancel", "")
+		cancel()
+		require.NoError(t, err)
+
+		<-ctx.Done()
+		countDocs := 0
+		for {
+			fetched, ok := <-fetchedChan
+			if !ok {
+				break
+			}
+
+			countDocs += len(fetched.Docs)
+		}
+		require.Less(t, countDocs, pgdb.FetchedChannelCapacity*pgdb.FetchedSliceCapacity+1)
+	})
+}
+
+func TestQueryDocumentsConnectionLost(t *testing.T) {
+	t.Parallel()
+
+	ctx := testutil.Ctx(t)
+
+	//pool := testutil.Pool(ctx, t, nil, zaptest.NewLogger(t))
+	pool, err := pgdb.NewPool(ctx, testutil.PoolConnString(t, nil), zaptest.NewLogger(t), false)
+	require.NoError(t, err)
+
+	dbName := testutil.Schema(ctx, t, pool)
+	collectionName := testutil.Table(ctx, t, pool, dbName)
+
+	for i := 1; i <= pgdb.FetchedChannelCapacity*pgdb.FetchedSliceCapacity*2; i++ {
+		require.NoError(t, pool.InsertDocument(
+			ctx, dbName, collectionName,
+			must.NotFail(types.NewDocument("id", fmt.Sprintf("%d", i))),
+		))
+	}
+
+	fetchedChan, err := pool.QueryDocuments(ctx, dbName, collectionName, "")
+	require.NoError(t, err)
+
+	<-fetchedChan
+	/*_, err = pool.Exec(ctx, "SELECT pg_terminate_backend(pg_backend_pid());")
```

Comment:
@AlekSi do you have any idea how to test this case? We wanted to emulate a problem with DB during iterations. Simple `pool.Close` doesn't work (it's hanging as the transaction is open), and also this `SELECT pg_terminate_backend(pg_backend_pid())` doesn't help either.

Reply:
I don't know. Let's add TODO to think about it in the future